internal/server/GameLoop.go

```diff
@@ -3,30 +3,115 @@ package server
+	// 2. pass to agents on each of the desired bikes a list of all agents trying to join
+	for bike, pendingAgents := range bikeRequests {
+		agents := s.megaBikes[bike].GetAgents()
+
+		responses := make([](map[uuid.UUID]bool), 0) // list containing all the agents' ranking
+		for _, agent := range agents {
+			responses = append(responses, agent.DecideJoining(pendingAgents))
+		}
+		// 3. accept agents based on the response outcome (it will have to be a ranking system, as only 8-n bikers can be accepted)
+		acceptedRanked := GetAcceptanceRanking(responses)
+		for _, accepted := range acceptedRanked[:(8 - len(s.megaBikes[bike].GetAgents()))] {
+			s.GetAgentMap()[accepted].ToggleOnBike()
+			s.SetBikerBike(s.GetAgentMap()[accepted], bike)
+		}
+	}
+
+	// map of the proposed lootboxes by bike (for each bike a list of lootbox proposals is made, with one lootbox proposed by each agent on the bike)
+	proposedDirections := make(map[uuid.UUID][]uuid.UUID)
+	// -------------------------------- DECIDE ACTION ---------------------------------
 	for agentId, agent := range s.GetAgentMap() {
 		fmt.Printf("Agent %s updating state \n", agentId)
 		agent.UpdateGameState(s)
 		agent.UpdateAgentInternalState()
+
 		switch agent.DecideAction() {
 		case objects.Pedal:
-			agent.DecideForce()
+			// --------------------- VOTING ROUTINE - STEP 1 --------------------------
+			// pitch proposal (desired lootbox)
+			bike := agent.GetBike()
+			if ids, ok := proposedDirections[bike]; ok {
+				proposedDirections[bike] = append(ids, agent.ProposeDirection())
+			} else {
+				proposedDirections[bike] = []uuid.UUID{agent.ProposeDirection()}
+			}
+
 		case objects.ChangeBike:
-			s.SetBikerBike(agent, agent.ChangeBike())
+			// decide which bike the agent is going to try and go to
+			// the bike id is set to be the desired bike and onbike is set to false
+			// so by looking at the values of onBike and megaBikeID it will be known
+			// whether the agent is trying to join a bike (and which one)
+
+			// the request is handled at the beginning of the next round, so the moving
+			// will only be finalised then
+			agent.SetBike(agent.ChangeBike())
+			agent.ToggleOnBike()
+
+			// the biker needs to be removed from the current bike as well
+			// it will be added to the desired one (if accepted) at the beginning of next loop
+			if oldBikeId, ok := s.megaBikeRiders[agent.GetID()]; ok {
+				s.megaBikes[oldBikeId].RemoveAgent(agent.GetID())
+			}
+
 		default:
 			panic("agent decided invalid action")
 		}
 	}
 
+	// pass the pitched directions of a bike to all agents on that bike and get their final vote
+	for bikeID, proposals := range proposedDirections {
+		// ----------------------------- VOTING ROUTINE - STEP 2 -----------------
+		// get second vote given everyone's proposal
+		// the finalVote can either be a ranking of proposed directions or a map from proposal to vote (between 0,1)
+		// we will try implementing both, the infrastructure should be the same
```

Comment:
I think you may have misinterpreted the voting mechanism.

The idea is that you don't supply a list of possible directions, which are then voted on. You instead elect a leader which has institutional power - it decides where the other agents go. Agents can defect however, so you can secretly have agents turning (i.e., not abstaining from turning), giving unintended steering (resulting in a change of leadership / ejection from bike / etc etc)

As such, you should consider having steer (a bearing) potentially equal to null, meaning that the agent doesn't interact. Then, the total steer is the average of all the results. If you like the leader (they're doing a good job) then you WILL abstain, since you want the best for the bike and trust in their abilities (we call this the emergence of experts / RRSI) 

We agreed that the pedalling is done as a whole - perhaps the leader suggests a power and everyone complies, maybe

Reply:
So for this round of iterations I think we are going to stick with the original plan as people have already started implementing their strategies based on it (see the PR comment for the voting structure). But I think this will be the goal of the next round of experiments